hs_access_control/models/user.py

```diff
@@ -2241,5 +2414,1036 @@ def can_undo_share(self, **kwargs):
+            assert isinstance(this_community, Community)
+
+        if not self.user.is_active:
+            raise PermissionDenied("Requesting user is not active")
+
+        return self.can_view_community(this_community)
+
+    def can_change_community_flags(self, this_community):
+        """
+        Whether the current user can change community flags:
+
+        :param this_community: community to query
+        :return: True if the user can set flags.
+
+        Usage:
+        ------
+
+            if my_user.can_change_community_flags(some_community):
+                some_community.active=False
+                some_community.save()
+
+        In practice:
+        ------------
+
+        This routine is called *both* when building views and when writing responders.
+        It should be called on both sides of the connection.
+
+            * In a view builder, it determines whether buttons are shown for flag changes.
+
+            * In a responder, it determines whether the request is valid.
+
+        At this point, the return value is synonymous with ownership or admin.
+        This may not always be true. So it is best to explicitly call this function
+        rather than assuming implications between functions.
+        """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_community, Community)
+
+        if not self.user.is_active:
+            raise PermissionDenied("Requesting user is not active")
+
+        return self.user.is_superuser or self.owns_community(this_community)
+
+    def can_delete_community(self, this_community):
+        """
+        Whether the current user can delete a community.
+
+        :param this_community: community to query
+        :return: True if the user can delete it.
+
+        Usage:
+        ------
+
+            if my_user.can_delete_community(some_community):
+                my_user.delete_community(some_community)
+            else:
+                raise PermissionDenied("Insufficient privilege")
+
+        In practice:
+        --------------
+
+        At this point, this is synonymous with ownership or admin. This may not always be true.
+        So it is best to explicitly call this function rather than assuming implications
+        between functions.
+        """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_community, Community)
+
+        if not self.user.is_active:
+            raise PermissionDenied("Requesting user is not active")
+
+        return self.user.is_superuser or self.owns_community(this_community)
+
+    @property
+    def communities(self):
+        """
+        return the communities of which a user is a member.
+
+        A user is a member of a community if the user is a member of one group in the community.
+        """
+        return Community.objects.filter(c2gcp__group__g2ugp__user=self.user)
```

Comment:
What if 
 - a 'group' to which the `self.user` belongs has been unshared with the community (GroupCommunityPrivilege: NONE)?
 - the `self.user` has been unshared with a `group` (UserGroupPrivilege:NONE) that is part of the community?

Reply:
Then the access control record will not even appear in the database. This was a change in version 2 of access control. No record will be matched except in the provenance tables.